### ISSUE TITLE
Self-closing <mutation> tag in XML will break the resulting 'if' block

### ISSUE THREAD

**Issue by Adam13531** [google/blockly#3664]
**Describe the bug**

If you provide toolbox XML with a self-closing mutation tag, the resulting 'if' block will be malformed.

**To Reproduce**

The easiest way to reproduce this is to apply this patch to the repo:

```diff
diff --git a/tests/playground.html b/tests/playground.html
index 3523caa6..6a07c912 100644
--- a/tests/playground.html
+++ b/tests/playground.html
@@ -692,6 +692,16 @@ var spaghettiXml = [
       <block type="logic_null" disabled="true"></block>
       <block type="logic_ternary"></block>
     </category>
+    <category name="Test" categorystyle="logic_category">
+      <block type="controls_if">
+        <mutation else="1"/>
+        <value name="IF0">
+          <block type="logic_boolean">
+            <field name="BOOL">TRUE</field>
+          </block>
+        </value>
+      </block>
+    </category>
     <category name="Loops" categorystyle="loop_category">
       <block type="controls_repeat_ext">
         <value name="TIMES">
```

Then, open `blockly/tests/playground.html` in the browser and click "Test" in the toolbox.

Expected: this screenshot that properly has the contents of the 'if' block:
![image](https://user-images.githubusercontent.com/7192897/73206785-6ba0f880-40f8-11ea-9a7f-66b396c999e1.png)

Actual outcome: this screenshot which shows a blank 'if' block:
![image](https://user-images.githubusercontent.com/7192897/73206804-73609d00-40f8-11ea-972e-7fecd7d513de.png)

**Desktop:**
 - OS: Windows 10
 - Browser: Chrome 79.0.3945.117, Firefox 72.0.1, and Safari 13.0.4 (15608.4.9.1.3) (i.e. I don't think this is browser- or OS-dependent)

**Stack Traces**

There is no error message in the console for this.

**Additional context**

I investigated for quite a while, but I can't definitively say what the problem is. What I *did* notice is that `domToBlockHeadless_` won't form any connections between blocks when the XML includes the self-closing tag.

Note that Blockly generates XML with self-closing mutation tags, which is the only way I even discovered this bug since I tried using my workspace's XML as a toolbox:

1. Configure an 'if' block with an 'else' mutation:
![image](https://user-images.githubusercontent.com/7192897/73207361-714b0e00-40f9-11ea-8eb9-1d8173e18ad2.png)

2. Click "Export to XML":

![image](https://user-images.githubusercontent.com/7192897/73207512-ac4d4180-40f9-11ea-9a93-f0eac39f7f68.png)

**Comment by moniika** [comment]
The issue with the block in the toolbox is that modern HTML does not permit self-closing tags and so the block xml:
```
<block type="controls_if">
  <mutation else="1"/>
    <value name="IF0">
    <block type="logic_boolean">
      <field name="BOOL">TRUE</field>
    </block>
  </value>
</block>
```
Is treated as though it was:
```
<block type="controls_if">
  <mutation else="1">
    <value name="IF0">
    <block type="logic_boolean">
      <field name="BOOL">TRUE</field>
    </block>
    </value>
  </mutation>
</block>
```
and so when html element is parsed, we lose the boolean value because it is nested inside the mutation instead of being a child of the block. This can be observed by inspecting the html.
Techncially, self-closing mutation tag does work if it is after the value, but otherwise there is issue.

**Comment by NeilFraser** [comment]
It's not just mutations, blocks are also exported to XML as self-closing.

```
  <block type="math_random_float" id="xskt0kOYqr$|I*VDr0i|" x="88" y="137"/>
```

That's not good.  I seem to remember adding a regexp that would un-self close generated XML.  But a cursory search doesn't find it in our codebase at the moment.

**Comment by samelhusseini** [comment]
Even though use XML for the toolbox, since it's injected in the page, the browser parses it with the HTML5 spec.
In HTML5, only [void elements](https://html.spec.whatwg.org/multipage/syntax.html#start-tags) are allowed to have self closing tags, otherwise it is considered an error and is parsed as if it wasn't there resulting in any following xml being treated as a child element of the closing node.

As @NeilFraser mentions, we shouldn't be exporting xml with self-closing tags.
In the similar vein, writing xml with self-closing tags in the toolbox will result in unexpected behaviour.

**Comment by samelhusseini** [comment]
@NeilFraser you might be referring to https://github.com/google/blockly/blob/master/demos/blockfactory/workspacefactory/wfactory_controller.js#L345

This does it in the block factory, but the same logic should probably be added to the xml generated by ``Blockly.Xml.domToText``.

**Comment by NeilFraser** [comment]
@samelhusseini Yes, that's exactly what I was thinking of.  Agreed, that regexp should be moved from block factory to core.

Normally I'd suggest a function in `Blockly.utils.xml`, but that file is intended to be swapped out with a different version for execution on Node.  Besides, `Blockly.Xml.domToText` already has one hilarious regexp, so adding a second would make them less lonely.

**Comment by mattbishop** [comment]
Hmm did the xml string generate from `domToPrettyText()`? It looks like the regex there is collapsing empty XML elements.
https://github.com/google/blockly/blob/master/core/xml.js#L335

**Comment by NeilFraser** [comment]
@mattbishop No, that code is innocent.  It is just converting:
```
<foo>
  <bar>
  </bar>
</foo>
```
to
```
<foo>
  <bar></bar>
</foo>
```
Otherwise the output is decidedly not pretty.

**Comment by samelhusseini** [comment]
Fixed in #3665

**Comment by Adam13531** [comment]
Great job, everyone! Thanks for the quick turnaround time. 😄
